test/helper/sending.bash

```diff
@@ -82,20 +132,35 @@ function _send_email() {
 # chosen. Sending more than one mail at any given point in time with this function
 # is UNDEFINED BEHAVIOR!
 function _send_email_and_get_id() {
-  [[ -v CONTAINER_NAME ]] || return 1
-
-  _wait_for_empty_mail_queue_in_container
-  _send_email "${@}"
-  _wait_for_empty_mail_queue_in_container
+  # Export the variable denoted by ${1} so everyone has access
+  export "${1:?Mail ID must be set for _send_email_and_get_id}"
+  # Get a "reference" to the content of the variable denoted by ${1} so we can manipulate the content
+  local -n ID_ENV_VAR_REF=${1:?}
+  # Prepare the message ID header here because will will shift away ${1} later
+  local MID="<${1}@dms-tests>"
+  # Get rid of ${1} so only the arguments for swaks remain
+  shift 1
 
-  local MAIL_ID
+  local QUEUE_ID MESSAGE_ID
   # The unique ID Postfix (and other services) use may be different in length
   # on different systems (e.g. amd64 (11) vs aarch64 (10)). Hence, we use a
   # range to safely capture it.
-  MAIL_ID=$(_exec_in_container tac /var/log/mail.log              \
-    | grep -E -m 1 'postfix/smtpd.*: [A-Z0-9]+: client=localhost' \
-    | grep -E -o '[A-Z0-9]{9,12}' || true)
+  local QUEUE_ID_REGEX='[A-Z0-9]{9,12}'
+
+  _wait_for_empty_mail_queue_in_container
+  local OUTPUT=$(_send_email "${@}" --header "Message-Id: ${MID}")
+  _wait_for_empty_mail_queue_in_container
+
+  # We store Postfix's queue ID first
+  ID_ENV_VAR_REF=$(_exec_in_container tac /var/log/mail.log                    \
+    | grep -E "postfix/cleanup.*: ${QUEUE_ID_REGEX}:.*message-id=${MID}" \
+    | grep -E --only-matching --max-count 1 "${QUEUE_ID_REGEX}" || :)
+  # But we also requre potential Dovecot sieve output, which requires the mesage ID,
+  # so we need to provide the message ID too.
+  ID_ENV_VAR_REF+="|${MID}"
```

Comment:
Is `$MID` not obvious to the caller of `_send_email_and_get_id()`? (`"<${1}@dms-tests>"`)

They could match against the Dovecot `msgid` with that same input then. Technically the whole need for the reference var vs just using a regular arg seems no longer required? 🤔 

I'm fine with the PR merging with whichever approach you want to use.

Reply:
> Is `$MID` not obvious to the caller of `_send_email_and_get_id()`? (`"<${1}@dms-tests>"`)

No, I actually think it is only obvious to me and you, but not to others.

> They could match against the Dovecot `msgid` with that same input then.

Still easier to just use `_print_mail_log_for_id` IMO.

> Technically the whole need for the reference var vs just using a regular arg seems no longer required? 🤔

I don't get why this is the case TBH. The reference is important to not rely on subshells, which hide exit codes and asserts.

> I'm fine with the PR merging with whichever approach you want to use.

I'd definitely go with refs, since there is very good reason for choosing them.

Reply:
> > Is `$MID` not obvious to the caller of `_send_email_and_get_id()`? (`"<${1}@dms-tests>"`)
> 
> No, I actually think it is only obvious to me and you, but not to others.

I meant, tests would just call something like:

```bash
# MID is set here, it's pretty obvious to the caller
# (who'd otherwise be able to pass this to current `_send_email_and_get_id()` anyway..)
local MID="<rspamd-gtube-reject@dms-tests>"

# Send an email as you normally would, but include the MID to match:
_send_email --expect-rejection \
  --header "Message-Id: ${MID}" \
  --body "XJS${GTUBE_SUFFIX}"

# Dedicated function that takes the MID, waits for empty queue and returns Queue ID
_print_mail_log_for_queue_id "$(_get_queue_id ${MID})"
# Actual asserts that matter:
assert_output --partial 'milter-reject'
assert_output --partial '5.7.1 Gtube pattern'

# Dovecot Sieve (spam to junk) shouldn't exist for this mail (QID doesn't exist on this log line)
_filter_service_log 'mail' "msgid=${MID}"
assert_failure
```

> > They could match against the Dovecot `msgid` with that same input then.
> 
> Still easier to just use `_print_mail_log_for_id` IMO.

As shown above with the Dovecot example, your method won't help as the queue ID is for Postfix, it's not part of the Dovecot log, unless we avoid providing our own message ID header and let Postfix generate one that includes the original queue ID (_this is not the queue ID that you necessarily use to match logs, as I have shown it can change while message ID persists_). 

Thus Queue ID and Message ID values won't necessarily cover all log lines of interest alone. Hence the logic to match by message ID, and get the queue ID of a specific log entry that it is known the queue ID can differ.

---

> > Technically the whole need for the reference var vs just using a regular arg seems no longer required? 🤔
> 
> I don't get why this is the case TBH. The reference is important to not rely on subshells, which hide exit codes and asserts.

> I'd definitely go with refs, since there is very good reason for choosing them.

I just prefer keeping the helpers simple. If the main purpose is just to get the Queue ID from a Message ID, I'm not sure how this differs from convention of other helpers regarding exit codes and asserts where the use of a reference is necessary.

If this was a nicer language like Rust I'd still probably opt for a function that takes an input parameter (as reference) and returns a value rather than a reference (_or mutating a value from a reference_).

I'm not great with bash, so if you feel it's really worthwhile/important, I'm not going to challenge that and trust your call, I just personally don't see the pragmatic benefit provided vs something simpler.

Reply:
> > > Is `$MID` not obvious to the caller of `_send_email_and_get_id()`? (`"<${1}@dms-tests>"`)
> > 
> > 
> > No, I actually think it is only obvious to me and you, but not to others.
> 
> I meant, tests would just call something like:
> 
> ```shell
> # MID is set here, it's pretty obvious to the caller
> # (who'd otherwise be able to pass this to current `_send_email_and_get_id()` anyway..)
> local MID="<rspamd-gtube-reject@dms-tests>"
> 
> # Send an email as you normally would, but include the MID to match:
> _send_email --expect-rejection \
>   --header "Message-Id: ${MID}" \
>   --body "XJS${GTUBE_SUFFIX}"
> 
> # Dedicated function that takes the MID, waits for empty queue and returns Queue ID
> _print_mail_log_for_queue_id "$(_get_queue_id ${MID})"
> # Actual asserts that matter:
> assert_output --partial 'milter-reject'
> assert_output --partial '5.7.1 Gtube pattern'
> 
> # Dovecot Sieve (spam to junk) shouldn't exist for this mail (QID doesn't exist on this log line)
> _filter_service_log 'mail' "msgid=${MID}"
> assert_failure
> ```

That seems more verbose than it needs to be, IMO. Moreover, with the current approach, building `MID` is easy if it is ever required. The line `_filter_service_log 'mail' "msgid=${MID}"` fails, but filtering for `${MID}` directly (as `_send_email_and_get_id` currently does) yields proper output.

> > > They could match against the Dovecot `msgid` with that same input then.
> > 
> > 
> > Still easier to just use `_print_mail_log_for_id` IMO.
> 
> As shown above with the Dovecot example, your method won't help as the queue ID is for Postfix, it's not part of the Dovecot log, unless we avoid providing our own message ID header and let Postfix generate one that includes the original queue ID (_this is not the queue ID that you necessarily use to match logs, as I have shown it can change while message ID persists_).
> 
> Thus Queue ID and Message ID values won't necessarily cover all log lines of interest alone. Hence the logic to match by message ID, and get the queue ID of a specific log entry that it is known the queue ID can differ.
> 
> > > Technically the whole need for the reference var vs just using a regular arg seems no longer required? 🤔
> > 
> > 
> > I don't get why this is the case TBH. The reference is important to not rely on subshells, which hide exit codes and asserts.
> 
> > I'd definitely go with refs, since there is very good reason for choosing them.
> 
> I just prefer keeping the helpers simple. If the main purpose is just to get the Queue ID from a Message ID, I'm not sure how this differs from convention of other helpers regarding exit codes and asserts where the use of a reference is necessary.
> 
> If this was a nicer language like Rust I'd still probably opt for a function that takes an input parameter (as reference) and returns a value rather than a reference (_or mutating a value from a reference_).
> 
> I'm not great with bash, so if you feel it's really worthwhile/important, I'm not going to challenge that and trust your call, I just personally don't see the pragmatic benefit provided vs something simpler.

I think complexity is not too high really for `_send_email_and_get_id`. I think the `local -n` reference is required and I'd definitely go for it :)

Reply:
> That seems more verbose than it needs to be, IMO. Moreover, with the current approach, building `MID` is easy if it is ever required.
> 
> The line `_filter_service_log 'mail' "msgid=${MID}"` fails, but filtering for `${MID}` directly (as `_send_email_and_get_id` currently does) yields proper output.

???

What is more verbose?
- The `_send_email` is the equivalent of your `_send_email_and_get_id --expect-rejection --body ...` for one of the rspamd calls. You pass a string to be part of the MID value instead of declaring the full header.
- Middle portion is your own test case
- Final part shows Dovecot `msgid` matching. In this test case the log matching is expected to fail as the assert afterwards implies. It doesn't exist with the rspamd body triggering reject.
- You would not get a match on this log line with your methods due to the MID being the search keyword needed, the queue ID isn't part of the log line 🤨 (_my comment mentioned all that_)

---

> I think complexity is not too high really for `_send_email_and_get_id`. I think the `local -n` reference is required and I'd definitely go for it :)

My example was to draw attention that it's doing extra work than appears necessary 🤷‍♂️ as opposed to something clearer which can also match a dovecot sieve `msgid` log line (_when one is present_).

Reply:
> > That seems more verbose than it needs to be, IMO. Moreover, with the current approach, building `MID` is easy if it is ever required.
> > The line `_filter_service_log 'mail' "msgid=${MID}"` fails, but filtering for `${MID}` directly (as `_send_email_and_get_id` currently does) yields proper output.
> 
> ???

Forget what I said, I was not reading correctly to begin with - sorry.

> What is more verbose?
> 
>   * The `_send_email` is the equivalent of your `_send_email_and_get_id --expect-rejection --body ...` for one of the rspamd calls. You pass a string to be part of the MID value instead of declaring the full header.
> 
>   * Middle portion is your own test case
> 
>   * Final part shows Dovecot `msgid` matching. In this test case the log matching is expected to fail as the assert afterwards implies. It doesn't exist with the rspamd body triggering reject.

Now I understand what you mean! We could implement this as well, I don't mind :)

>   * You would not get a match on this log line with your methods due to the MID being the search keyword needed, the queue ID isn't part of the log line 🤨 (_my comment mentioned all that_)

I don't quite get that though; `${MID}` is part of the ID we're currently using (second part of the regex (`|${MID}`)). I think we're talking about different things here; which log line are you referring to with "this"?

> > I think complexity is not too high really for `_send_email_and_get_id`. I think the `local -n` reference is required and I'd definitely go for it :)
> 
> My example was to draw attention that it's doing extra work than appears necessary 🤷‍♂️ as opposed to something clearer which can also match a dovecot sieve `msgid` log line (_when one is present_).

The `rspamd_full` test shows that `_send_email_and_get_id` is already capable of providing Dovecot Sieve log lines, or am I missing something?

https://github.com/docker-mailserver/docker-mailserver/blob/06fab3f12964b9ce5d5d6e889de28598635c6cf5/test/tests/parallel/set1/spam_virus/rspamd_full.bats#L225-L226